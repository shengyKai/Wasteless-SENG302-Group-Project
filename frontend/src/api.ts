/*
 * Created on Wed Feb 10 2021
 *
 * The Unlicense
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or distribute
 * this software, either in source code form or as a compiled binary, for any
 * purpose, commercial or non-commercial, and by any means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors of this
 * software dedicate any and all copyright interest in the software to the public
 * domain. We make this dedication for the benefit of the public at large and to
 * the detriment of our heirs and successors. We intend this dedication to be an
 * overt act of relinquishment in perpetuity of all present and future rights to
 * this software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 */

/**
 * Declare all available services here
 */
import axios from 'axios';
const SERVER_URL = process.env.VUE_APP_SERVER_ADD;

const instance = axios.create({
  baseURL: SERVER_URL,
  timeout: 2000,
  withCredentials: true,
});

type MaybeError<T> = T | string;

export type User = {
  id: number,
  firstName: string,
  lastName: string,
  middleName?: string,
  nickname?: string,
  bio?: string,
  email: string,
  dateOfBirth: string,
  phoneNumber?: string,
  homeAddress: Location,
  created?: string,
  role?: "user" | "globalApplicationAdmin" | "defaultGlobalApplicationAdmin",
  businessesAdministered?: Business[],
};

export type Location = {
  streetNumber?: string,
  streetName?: string,
  city?: string,
  region?: string,
  country: string,
  postcode?: string,
};

export type CreateUser = {
  firstName: string,
  lastName: string,
  middleName?: string,
  nickname?: string,
  bio?: string,
  email: string,
  dateOfBirth: string,
  phoneNumber?: string,
  homeAddress: Location,
  password: string,
};

export type BusinessType = 'Accommodation and Food Services' | 'Retail Trade' | 'Charitable organisation' | 'Non-profit organisation';

export const BUSINESS_TYPES: BusinessType[] = ['Accommodation and Food Services', 'Retail Trade', 'Charitable organisation', 'Non-profit organisation'];

export type Business = {
  id: number,
  administrators?: User[],
  name: string,
  description?: string,
  address: string,
  businessType: BusinessType,
  created?: string,
};

export type CreateBusiness = {
  name: string,
  description?: string,
  address: string,
  businessType: BusinessType,
};

export type Image = {
  id: number,
  filename: string,
  thumbnailFilename: string,
};

export type Product = {
  id: string,
  name: string,
  description?: string,
  manufacturer?: string,
  recommendedRetailPrice?: number,
  created?: string,
  images: Image[],
};

export type CreateProduct = Omit<Product, 'created' | 'images'>;

function isLocation(obj: any): obj is Location {
  if (obj === null || typeof obj !== 'object') return false;
  if (obj.streetNumber !== undefined && typeof obj.streetNumber !== 'string') return false;
  if (obj.streetName !== undefined && typeof obj.streetName !== 'string') return false;
  if (obj.city !== undefined && typeof obj.city !== 'string') return false;
  if (obj.region !== undefined && typeof obj.region !== 'string') return false;
  if (typeof obj.country !== 'string') return false;
  if (obj.postcode !== undefined && typeof obj.postcode !== 'string') return false;

  return true;
}

function isUser(obj: any): obj is User {
  if (obj === null || typeof obj !== 'object') return false;
  if (typeof obj.id !== 'number') return false;
  if (typeof obj.firstName !== 'string') return false;
  if (typeof obj.lastName !== 'string') return false;
  if (obj.middleName !== undefined && typeof obj.middleName !== 'string') return false;
  if (obj.nickname !== undefined && typeof obj.nickname !== 'string') return false;
  if (obj.bio !== undefined && typeof obj.bio !== 'string') return false;
  if (typeof obj.email !== 'string') return false;
  if (typeof obj.dateOfBirth !== 'string') return false;
  if (obj.phoneNumber !== undefined && typeof obj.phoneNumber !== 'string') return false;
  if (!isLocation(obj.homeAddress)) return false;
  if (obj.created !== undefined && typeof obj.created !== 'string') return false;
  if (obj.role !== undefined && !['user', 'globalApplicationAdmin', 'defaultGlobalApplicationAdmin'].includes(obj.role))
    return false;
  if (obj.businessesAdministered !== undefined && !isBusinessArray(obj.businessesAdministered)) return false;

  return true;
}

function isBusiness(obj: any): obj is Business {
  if (obj === null || typeof obj !== 'object') return false;
  if (typeof obj.id !== 'number') return false;
  if (obj.administrators !== undefined && !isUserArray(obj.administrators)) return false;
  if (typeof obj.name !== 'string') return false;
  if (obj.description !== undefined && typeof obj.description !== 'string') return false;
  if (typeof obj.address !== 'object') return false;
  if (!BUSINESS_TYPES.includes(obj.businessType)) return false;
  if (obj.created !== undefined && typeof obj.created !== 'string') return false;

  return true;
}

function isNumberArray(obj: any): obj is number[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (typeof elem !== 'number') return false;
  }
  return true;
}

function isUserArray(obj: any): obj is User[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (!isUser(elem)) return false;
  }
  return true;
}

function isBusinessArray(obj: any): obj is Business[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (!isBusiness(elem)) return false;
  }
  return true;
}

type OrderBy = 'userId' | 'relevance' | 'firstName' | 'middleName' | 'lastName' | 'nickname' | 'email' | 'address';

/**
 * Sends a search query to the backend.
 *
 * @param query Query string to search for
 * @param pageIndex Index of page to start the results from (1 = first page)
 * @param resultsPerPage Number of results to return per page
 * @param orderBy Specifies the method used to sort the results
 * @param reverse Specifies whether to reverse the search results (default order is descending for relevance and ascending for all other orders)
 * @returns List of user infos for the current page or an error message
 */
export async function search(query: string, pageIndex: number, resultsPerPage: number, orderBy: OrderBy, reverse: boolean): Promise<MaybeError<User[]>> {
  let response;
  try {
    response = await instance.get('/users/search', {
      params: {
        searchQuery: query,
        page: pageIndex,
        resultsPerPage,
        orderBy,
        reverse: reverse.toString(),
      }
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    return `Request failed: ${status}`;
  }

  if (!isUserArray(response.data)) {
    return 'Response is not user array';
  }

  return response.data;
}

/**
 * Sends a query for the number of search results for a given query to the backend.
 *
 * @param query Query string to search for
 * @returns Number of search results or an error message
 */
export async function getSearchCount(query: string): Promise<MaybeError<number>> {
  let response;
  try {
    response = await instance.get('/users/search/count', {
      params: {
        searchQuery: query,
      }
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    return `Request failed: ${status}`;
  }

  if (typeof response.data?.count !== 'number') {
    return 'Response is not number';
  }

  return response.data.count;
}

/**
 * Queries the backend for a specific user by their id.
 *
 * @param id User id, if ommitted then fetch the logged in user's info
 * @returns User info for the given id or an error message
 */
export async function getUser(id: number): Promise<MaybeError<User>> {
  let response;
  try {
    response = await instance.get('/users/' + id);
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    return `Request failed: ${status}`;
  }

  if (!isUser(response.data)) {
    return 'Response is not user';
  }

  return response.data;
}

/**
 * Logs in to the given user account by setting the authentication cookie.
 *
 * @param email User email
 * @param password User password
 * @returns The now logged in user ID if operation is successful, otherwise a string error.
 */
export async function login(email: string, password: string): Promise<MaybeError<number>> {
  let response;
  try {
    response = await instance.post('/login', {
      email: email,
      password: password,
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid credentials';
    return `Request failed: ${status}`;
  }
  let id = response.data.userId;
  if (typeof id !== 'number') return 'Invalid response';

  return id;
}

/**
 * Creates a user with the given properties.
 * Note: This doesn't automatically log the user in.
 *
 * @param user Initial user properties
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function createUser(user: CreateUser): Promise<MaybeError<undefined>> {
  try {
    await instance.post('/users', user);
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid create user request';
    if (status === 409) return 'Email in use';
    return 'Request failed: ' + status;
  }

  return undefined;
}

/**
 * Makes the given user a GAA, if the current user is a DGAA.
 *
 * @param userId User to make GAA
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function makeAdmin(userId: number): Promise<MaybeError<undefined>> {
  try {
    await instance.post(`/users/${userId}/makeAdmin`);
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 403) return 'Operation not permitted';
    if (status === 406) return 'User does not exist';

    return 'Request failed: ' + status;
  }
  return undefined;
}

/**
 * Revokes the given user's admin rights
 *
 * @param userId User to revoke permissions
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function revokeAdmin(userId: number): Promise<MaybeError<undefined>> {
  try {
    await instance.post(`/users/${userId}/revokeAdmin`);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 403) return 'Operation not permitted';
    if (status === 406) return 'User does not exist';

    return 'Request failed: ' + status;
  }
  return undefined;
}

/**
 * Creates a business
 *
 * @param business The properties to create the business with
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function createBusiness(business: CreateBusiness): Promise<MaybeError<undefined>> {
  try {
    await instance.post('/businesses', business);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';

    return 'Request failed: ' + status;
  }

  return undefined;
}

/**
 * Add a product to a businesses catalogue.
 *
 * @param businessId The business to add the product to
 * @param product The properties to create a product with
 * @return undefined if operation is successful, otherwise a string error
 */
export async function createProduct(businessId: number, product: CreateProduct): Promise<MaybeError<undefined>> {
  try {
    await  instance.post(`/businesses/${businessId}/products`, product);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 403) return 'Operation not permitted';

    if (status === 400) {
      // TODO Not sure exactly how the backend is going to communicate with us that the product code
      // is unavailable.

      // eslint-disable-next-line no-constant-condition
      if (false) {
        return 'Product code unavailable';
      }

      return 'Invalid parameters';
    }

    return 'Request failed: ' + status;
  }
  return undefined;
}

/**
 * Fetches a business with the given id.
 *
 * @param businessId Business id to fetch
 * @returns The requested business if operation is successful, otherwise a string error.
 */
export async function getBusiness(businessId: number): Promise<MaybeError<Business>> {
  let response;
  try {
    response = await instance.get(`/businesses/${businessId}`);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 406) return 'Business not found';

    return 'Request failed: ' + status;
  }

  if (!isBusiness(response.data)) {
    return 'Invalid response type';
  }

  return response.data;
}