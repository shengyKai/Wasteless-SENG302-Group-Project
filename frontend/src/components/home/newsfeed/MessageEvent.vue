<template>
  <!--<Event :event="event" :title="title">-->
  <div>
    <v-card-text v-bind:style="{ height: computedHeight }">
      New message from: {{userName}}
      Regarding: "{{cardTitle}}" in the {{section}} section of the Marketplace
      {{message}}
    </v-card-text>
    <div class="foot-tools">
      <v-icon v-if="expanded" @click="expand" title="Collapse text">mdi-arrow-up-drop-circle</v-icon>
      <v-icon v-else @click="expand" title="Expand text">mdi-arrow-down-drop-circle</v-icon>
      <v-btn text color="primary" @click="sendReply">reply</v-btn>
    </div>
  </div>
  <!--</Event>-->
</template>

<script>
//import Event from './Event';

export default {
  name: "MessageEvent",
  components: {
    //Event
  },
  props: {
    //event: {
    //  type: Object,
    //  required: true,
    //},
  },
  data() {
    return {
      title: "New message",
      expanded: false,
      computedHeight: '200em'
    };
  },
  methods: {
    expand() {
      this.expanded = !this.expanded;
      if (this.expanded) {
        this.computedHeight = undefined;
      } else {
        this.computedHeight = '200em';
      }
    },
    sendReply() {
      // Open form to write message back
    }
  },
  computed: {
    // Fluff placeholder
    userName() {
      // First name + Last name + ?(Nickname)
      return "Test Name";
    },
    cardTitle() {
      return "Hungry for apples?";
    },
    section() {
      return "For Sale";
    },
    message() {
      return "Voluptas quaerat consequatur esse nihil ipsa sed natus sed. Autem et atque pariatur optio porro. Ex odio " +
          "perferendis voluptas nihil suscipit earum at et. Ab eum corporis deleniti. Repudiandae quaerat totam numquam " +
          "quis iusto tempore. Harum omnis totam ut.  Commodi quibusdam quo provident temporibus. Ut est porro mollitia " +
          "est quidem magnam ex. Modi accusantium beatae quas aut.";
    }
  }
};
</script>

<style scoped>
.foot-tools {
  display: flex;
  flex-wrap: wrap;
}
</style>