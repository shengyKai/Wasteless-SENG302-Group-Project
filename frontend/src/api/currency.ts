import { MaybeError } from "@/api/internal";

/**
 * Currency information for a specific country.
 */
export type Currency = {
  code: string,
  name: string,
  symbol: string
};

type CurrencyOrError = Currency | { errorMessage: string }

/**
 * An object which only has the attribute 'currencies', which is a list of Currency objects. The API response is expected
 * to contain an array of objects of this type
 */
type CurrenciesContainer = {
  currencies: Currency[];
};

/**
 * Determines if the given object is an instance of Currency.
 * @param obj The object to be checked.
 * @returns True if the object has the attributes of a Currency object, false otherwise.
 */
function isCurrency(obj: any): obj is Currency {
  if (obj === null || typeof obj !== 'object') return false;
  if (typeof obj.code !== 'string') return false;
  if (typeof obj.name !== 'string') return false;
  if (typeof obj.symbol !== 'string') return false;
  return true;
}

/**
 * Determines if the given object is an array of object of type Currency.
 * @param obj The object to be checked.
 * @returns True if the object is an array where all elements are Currency objects, false otherwise.
 */
function isCurrencyArray(obj: any): obj is Currency[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (!isCurrency(elem)) return false;
  }
  return true;
}

/**
 * Check that the body of the API response has the expected format. It should be an array of length 1, where they
 * entry in the array is a JSON with a "currencies" field, and that field contains an array of currency objects.
 * @param response The body of the response received from the API.
 * @returns True if the response is in the expected format, false otherwise.
 */
function currencyResponseHasExpectedFormat(response: any): response is [CurrenciesContainer] {
  if (!Array.isArray(response)) return false;
  if (response.length !== 1) return false;
  const country = response[0];
  if (!isCurrencyArray(country.currencies)) return false;
  return true;
}

/**
 * Make a request to the RESTCounties API to find the currency associated with the given country name.
 * If the request is successful the currency from the API will be returned. If it is unsuccessful then
 * a default currency object with blank code, name and symbol fields will be returned.
 * @param country The name of a country to use in the API request for the currency.
 * @returns A list containing the Currency of the country(defaultCurrency if any errors occur) at index 0,
 * and an error message at index 1 if any errors occur.
 */
export async function currencyFromCountry(country: string): Promise<CurrencyOrError> {

  const response = await queryCurrencyAPI(country);

  if (typeof response === 'string') {
    console.warn(response);
    return { errorMessage: response };
  }

  const currency = await getCurrencyFromAPIResponse(response);

  if (typeof currency === 'string') {
    console.warn(currency);
    return { errorMessage: currency };
  }

  return currency;
}

/**
 * This method takes a string with the name of a country and queries the RESTCountries API to find the
 * currency associated with that country. The API's response will be returned if a response with status
 * code 200 is received, otherwise an error message will be returned.
 * @param country The name of the country to query the API for.
 * @return the response received from the RESTCounties API or a string error message.
 */
export async function queryCurrencyAPI(country: string): Promise<MaybeError<Response>> {

  const queryUrl = `https://restcountries.eu/rest/v2/name/${country}?fullText=true&fields=currencies`;

  const response = await fetch(queryUrl)
    .catch(error => {
      return `Failed to reach ${queryUrl}`;
    });
  if (typeof response === 'string') {
    return response;
  }
  if (response.status === 404) {
    return `No country with name ${country} was found`;
  }
  if (response.status !== 200) {
    return `Request failed: ${response.status}`;
  }
  return response;
}

/**
 * This method checks the format of the API response and extracts a currency object from the JSON body
 * if the response format is correct. If it is not correct then an error message is returned.
 * @param response A currency object extracted from the response body or an error message.
 */
export async function getCurrencyFromAPIResponse(response: Response): Promise<MaybeError<Currency>> {
  const responseBody = await response.json();

  if (!currencyResponseHasExpectedFormat(responseBody)) {
    return 'API response was not in readable format';
  }

  return responseBody[0].currencies[0];
}