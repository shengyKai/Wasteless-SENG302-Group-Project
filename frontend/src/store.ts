import {User,  Business, getUser, login} from './api';
import Vuex, { Store, StoreOptions } from 'vuex';
import { COOKIE, deleteCookie, setCookie } from './utils';

export type StoreData = {
  user: User | null,
  activeRole: { type: "user" | "business", id: number} | null,
};

export function createOptions(): StoreOptions<StoreData> {
  return {
    state: {
      user: null,
      activeRole: null,
    },
    mutations: {
      setUser (state, payload: User) {
        state.user = payload;
        // If the payload contains a user ID, user is now logged in. Set their session cookie.
        if (payload.id) {
          deleteCookie(COOKIE.USER.toUpperCase());
          deleteCookie(COOKIE.USER.toLowerCase());
          setCookie(COOKIE.USER, payload.id);
        }
      },
      logoutUser (state) {
        state.user = null;
        deleteCookie(COOKIE.USER);
      }
    },
    getters: {
      isLoggedIn (state) {
        return state.user !== null;
      },
      role (state) {
        return state.user?.role;
      }
    },
    actions: {
      async login(context, { email, password }) {
        let userId = await login(email, password);
        if (typeof userId === 'string') {
          console.warn(userId);
          return;
        }
        let user = await getUser(userId);
        if (typeof user === 'string') {
          console.error(user);
          return;
        }
        context.commit('setUser', user);
      }
    }
  };
}

export default function createStore() {
  return new Vuex.Store(createOptions());
}
