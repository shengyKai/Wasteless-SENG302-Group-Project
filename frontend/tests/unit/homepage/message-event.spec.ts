
import Vue from 'vue';
import Vuetify from 'vuetify';
import { createLocalVue, mount, Wrapper } from '@vue/test-utils';
import MessageEvent from '@/components/home/newsfeed/MessageEvent.vue';
import * as api from '@/api/internal';

import Vuex, { Store } from 'vuex';
import { getStore, resetStoreForTesting, StoreData } from '@/store';
import { castMock, makeTestUser } from '../utils';

Vue.use(Vuetify);

jest.mock('@/api/internal', () => ({
  messageConversation: jest.fn(),
  getMessagesInConversation: jest.fn(),
}));

const messageConversation = castMock(api.messageConversation);
const getMessagesInConversation = castMock(api.getMessagesInConversation);

const sellerUser = makeTestUser(100);
const buyerUser = makeTestUser(50);

const testCard: api.MarketplaceCard = {
  id: 1,
  creator: sellerUser,
  section: 'ForSale',
  created: "01/01/2020",
  lastRenewed: "02/01/2020",
  displayPeriodEnd: "05/01/2020",
  title: "test_title",
  description: "test_desription",
  keywords: []
};

describe('MessageEvent.vue', () => {
  let wrapper: Wrapper<any>;
  let eventWrapper: Wrapper<any>;
  let vuetify: Vuetify;
  // The global store to be used
  let store: Store<StoreData>;

  beforeEach(async () => {
    const localVue = createLocalVue();
    vuetify = new Vuetify();

    localVue.use(Vuex);
    resetStoreForTesting();
    store = getStore();
    store.state.user = buyerUser;

    wrapper = mount(MessageEvent, {
      localVue,
      vuetify,
      store,
      propsData: {
        event: {
          type: 'MessageEvent',
          id: 7,
          tag: 'none',
          created: '2021-01-02T11:00:00Z',
          message: {
            id: 9,
            senderId: buyerUser.id,
            created: '2021-01-02T11:00:00Z',
            content: 'test message 1',
          },
          conversation: {
            card: testCard,
            buyer: buyerUser,
            id: 3,
          },
          participantType: 'seller',
        },
      }
    });
    eventWrapper = wrapper.findComponent({name: 'Event'});
  });

  /**
   * Finds the button that controls sending a new message
   * @returns A wrapper around the send button
   */
  function findSendButton() {
    const buttons = wrapper.findAllComponents({ name: 'v-btn' });
    const filtered = buttons.filter(button => button.text().includes('Send'));
    expect(filtered.length).toBe(1);
    return filtered.at(0);
  }

  /**
   * Finds the button that controls loading more messages
   * @returns A wrapper around the load more button
   */
  function findLoadMoreButton() {
    const buttons = wrapper.findAllComponents({ name: 'v-btn' });
    const filtered = buttons.filter(button => button.text().includes('Load more'));
    expect(filtered.length).toBe(1);
    return filtered.at(0);
  }

  it('If new message is from other participant then title should be "New message from..."', () => {
    expect(eventWrapper.vm.title).toBe('Conversation with ' + sellerUser.firstName);
  });

  it('If new message is from self participant then title should be "Conversation with..."', async () => {
    await wrapper.setData({
      event: {
        message: {
          senderId: sellerUser.id,
        },
      },
    });
    await Vue.nextTick();
    expect(eventWrapper.vm.title).toBe('New message from: ' + sellerUser.firstName);
  });

  it('If no message is provided then the send button is enabled until a message tries to be sent', async () => {
    expect(findSendButton().props().disabled).toBeFalsy();

    await findSendButton().trigger('click');
    await Vue.nextTick();

    expect(findSendButton().props().disabled).toBeTruthy();
    expect(messageConversation).not.toBeCalled();
  });

  it('If the message is too long then the send button is disabled', async () => {
    await wrapper.setData({
      directMessageContent: 'a'.repeat(201),
    });
    await Vue.nextTick();
    expect(findSendButton().props().disabled).toBeTruthy();
  });

  it('If a message is provided then the send button is enabled', async () => {
    await wrapper.setData({
      directMessageContent: 'This is a message',
    });
    await Vue.nextTick();
    expect(findSendButton().props().disabled).toBeFalsy();
  });

  it('If the send button is pressed then the message should be sent', async () => {
    messageConversation.mockResolvedValueOnce(undefined);

    await wrapper.setData({
      directMessageContent: 'This is a message',
    });
    await Vue.nextTick();
    await findSendButton().trigger('click');
    await Vue.nextTick();

    expect(messageConversation).toBeCalledWith(testCard.id, buyerUser.id, buyerUser.id, 'This is a message');
    expect(eventWrapper.props().error).toBe(undefined); // No error is shown
    expect(wrapper.vm.directMessageContent).toBe(''); // Textbox is cleared
  });


  it('If the sending a message results in an error then it should be shown', async () => {
    messageConversation.mockResolvedValueOnce('test_error_message');

    await wrapper.setData({
      directMessageContent: 'This is a message',
    });
    await Vue.nextTick();
    await findSendButton().trigger('click');
    await Vue.nextTick();

    expect(eventWrapper.props().error).toBe('test_error_message');
    expect(wrapper.vm.directMessageContent).toBe('This is a message'); // Textbox is not cleared
  });

  it('Initial message should be displayed', () => {
    expect(wrapper.text()).toContain('test message 1');
  });

  it('Clicking load more will request more messages and displayed', async () => {
    getMessagesInConversation.mockResolvedValueOnce({
      count: 100,
      results: [
        {
          id: 6,
          senderId: buyerUser.id,
          created: new Date().toString(),
          content: 'test message 1',
        },
        {
          id: 4,
          senderId: buyerUser.id,
          created: new Date().toString(),
          content: 'test message 2',
        },
      ],
    });

    await findLoadMoreButton().trigger('click');

    await Vue.nextTick();

    expect(getMessagesInConversation).toBeCalledWith(testCard.id, buyerUser.id, 1, 1 + 10);
    expect(eventWrapper.props().error).toBe(undefined); // No error is shown

    expect(wrapper.text()).toContain('test message 1');
    expect(wrapper.text()).toContain('test message 2');
  });

  it('If load more results in an error then it is shown', async () => {
    getMessagesInConversation.mockResolvedValueOnce('test_error_message');

    await findLoadMoreButton().trigger('click');
    await Vue.nextTick();

    expect(eventWrapper.props().error).toBe('test_error_message');
  });

  it('If load more does not return any extra results, then the load more button is hidden', async () => {
    getMessagesInConversation.mockResolvedValueOnce({
      count: 1,
      results: [{
        id: 9,
        senderId: buyerUser.id,
        created: '2021-01-02T11:00:00Z',
        content: 'test message 1',
      }],
    });

    await findLoadMoreButton().trigger('click');
    await Vue.nextTick();

    const buttons = wrapper.findAllComponents({ name: 'v-btn' });
    const filtered = buttons.filter(button => button.text().includes('Load more'));
    expect(filtered.length).toBe(0);
  });

  it('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
