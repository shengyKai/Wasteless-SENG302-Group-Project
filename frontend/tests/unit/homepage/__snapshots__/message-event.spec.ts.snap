// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageEvent.vue Matches snapshot 1`] = `
<div>
  <div class="delete">
    <div class="float=right ma-1"><button type="button" class="v-icon notranslate deleteButton v-icon--link mdi mdi-trash-can theme--light red--text"></button></div>
    <div class="v-card__title">
      Conversation with test_firstname100
    </div>
    <div class="v-card__subtitle">
      03 Jan 2021, 00:00
    </div>
    <div class="v-card__text pb-1"><strong>
      Regarding: test_title
    </strong>
      <div class="overflow-y-auto d-flex flex-column-reverse pa-1 rounded" style="max-height: 350px; width: 100%;">
        <div class="message align-self-end">
          <div class="elevation-2 pa-1 rounded-lg mt-1 message-self">
            test message 1
          </div><span class="v-tooltip v-tooltip--bottom"><!----></span>
        </div>
        <div class="align-self-center"><button type="button" class="v-btn v-btn--has-bg v-btn--rounded theme--light v-size--small secondary"><span class="v-btn__content">Load more</span></button></div>
      </div>
    </div>
    <form novalidate="novalidate" class="v-form">
      <div class="row ma-0 mb-n3">
        <div class="pr-0 py-0 col">
          <div class="v-input v-textarea v-textarea--auto-grow v-textarea--no-resize theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-input__prepend-inner">
                  <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate mdi mdi-comment theme--light"></i></div>
                </div>
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span>â€‹</span></legend>
                </fieldset>
                <div class="v-text-field__slot"><label for="input-374" class="v-label theme--light" style="left: 0px; position: absolute;">Enter Message</label><textarea id="input-374" rows="1"></textarea></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
                <div class="v-counter theme--light">0 / 200</div>
              </div>
            </div>
          </div>
        </div>
        <div class="py-0 col col-auto"><button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary"><span class="v-btn__content">
          Send
        </span></button></div>
      </div>
    </form>
    <!---->
    <div class="row">
      <div class="col"><span class="ma-2 ml-4 mb-3 v-chip v-chip--clickable v-chip--label theme--light v-size--default none white--text" aria-haspopup="true" aria-expanded="false"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i>
              Tag
            </span></span><span class="v-tooltip v-tooltip--bottom"><!----></span></div>
    </div>
    <div class="row">
      <div class="col">
        <transition-stub name="expand-transition">
          <div class="mx-auto v-card v-sheet theme--light" style="max-width: 1200px; display: none;">
            <div class="font-weight-medium"><span class="ml-4">
                Change your Tag:
              </span></div>
            <div class="ml-3"><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default none white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default red white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default orange white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default yellow white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default green white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default blue white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default purple white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span></div>
          </div>
        </transition-stub>
      </div>
    </div>
  </div>
</div>
`;
